modin/core/dataframe/pandas/partitioning/partition.py

```diff
@@ -51,6 +51,21 @@ def get(self):
         """
         pass
 
+    @property
+    def list_of_blocks(self):
```

Comment:
Why did we decide to use `list_of_blocks` for a block partition? Doesn't it seem confusing?

Reply:
@prutskov, @vnlitvinov , and I thought it was more confusing to have `list_of_blocks` for non-full-axis virtual partitions, and a separate accessor like `block` or `data` for block partitions. non-full-axis virtual partitions and block partitions are also supposed to expose the same API.

In the long term, I want to consolidate the block and virtual partitions into a single class for each engine and remove the `full_axis` param from the axis partition constructor. We would always use the `num_splits` argument to `apply` explicitly to specify the number of splits. That will reduce code duplication and prevent bugs like the many we've had from not treating non-full-axis virtual partitions the same as full-axis virtual partitions, e.g. #4691, #4692, #4676, #4464. Using a common `list_of_blocks` for both partition class takes us toward that goal.

One alternative was to make all partition classes have a block accessor like `data()` that forces materialization of virtual partitions with multiple blocks, but this didn't seem necessary for any existing use case, and it seemed better to combine the blocks within the remote `apply` call for cases like binary operations, where we [use all the blocks](https://github.com/modin-project/modin/blob/eee5f435f68786778184c4886ff25d386fce0c4f/modin/core/dataframe/pandas/partitioning/partition_manager.py#L1288).

Reply:
> it was more confusing to have `list_of_blocks` for non-full-axis virtual partitions

I do not understand this. A virtual partition can encompass multiple block partitions. By calling `list_of_blocks` we would get the partitions.

> it was more confusing to have a separate accessor like block or data for block partitions

I do not understand this too. By calling partition.data we would get the data.

> non-full-axis virtual partitions and block partitions are also supposed to expose the same API.

Not sure that will lead to a nice design but will be happy to look at suggestions.

Reply:
I have below some inline responses to parts of your question. There was a request for comments (RFC) for this PR, but it was not public until today. [Here it is](https://www.notion.so/ponderdata/Unify-access-to-physical-data-of-partitions-93e06826ebe14937b3bc722c845947ee).

> I do not understand this too. By calling [partition.data](http://partition.data/) we would get the data.

If you can access the "updates" in the RFC link, you can scroll down to my comment starting with "still think list_of_blocks is better" and see the comment thread there. That section will hopefully explain more clearly the rationale for using list_of_blocks. Unfortunately you have to dig for the comment thread like this because we deleted the text that the comments were originally referring to.

> Not sure that will lead to a nice design but will be happy to look at suggestions.

Having non-full-axis virtual partitions and block partition expose the same API is not a proposal. It's already an assumption in the code that uses partitions, and it was so even prior to this PR. virtual partitions with `full_axis=False` are supposed to behave like block partitions, e.g. calling `apply` on them returns a single partition instead of a list of partitions.

> In the long term, I want to consolidate the block and virtual partitions into a single class 

I'll make sure to send a public proposal about this to you and other active modin maintainers before trying to implement it.

Reply:
Unfortunately, I can't access the RFC. If you could provide me with some brief comment, that would be great.

> I'll make sure to send a public proposal about this to you and other active modin maintainers before trying to implement it.

Thanks!